File: CppSolution/RemoveNthNodeFromEndofList.cpp
#include <iostream>
using namespace std;


struct ListNode 
{
	int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) 
    {
        ListNode *slow = head;
        ListNode *fast = head;
        int count = 0;
        while (count != n)
        {
            fast = fast->next;
            count++;
        }

        if (fast == NULL)
        {
            head = head->next;
            return head;
        }
        while (fast->next != NULL)
        {
            slow = slow->next;
            fast = fast->next;
        }
        slow->next = slow->next->next;
        return head;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/RemoveNthNodeFromEndofList.java
/**
 * Created by xyunpeng on 3/20/16.
 */
public class RemoveNthNodeFromEndofList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode walker = dummy;
        ListNode runner = dummy;
        int i = 0;
        while (i <= n && runner != null) {
            runner = runner.next;
            i++;
        }

        while (runner != null) {
            runner = runner.next;
            walker = walker.next;
        }

        if (walker.next != null) {
            walker.next = walker.next.next;
        }

        return dummy.next;
    }
}
